---
title: "CUBI advanced R bioinformatics course / 01"
author: ""
date: "`r Sys.Date()`"
output:
  ioslides_presentation: default
abstract: |

toc: no
---

```{r,echo=FALSE}
## Set default options for the knitr RMD processing
knitr::opts_chunk$set(echo=FALSE,warning=FALSE,message=FALSE,fig.width=5,fig.height=5,cache=TRUE,autodep=TRUE, results="hide")
```

```{r libraries,cache=FALSE}
```

# CUBI advanced R bioinformatics course

## Data

January Weiner <january.weiner@bihealth.de>

## Course files

Course repository: https://github.com/bihealth/Rcrashcourse

To view the lectures, download the HTML file and open it in a browser.

## Aims

 * Empower you in collaboration with CUBI
 * Explain what you see in our reports
 * Give you the power to create own analyses and beautiful illustrations

# Topics

## Linear models and co

 * Basic analysis of differential transcription
 * Interactions
 * Contrasts

## Functional analysis

 * Gene set enrichments
 * Visualizations
 * Investigation 

## Gene IDs and co.

 * org.XX.eg.db
 * Orthology:
   * BiomaRt
   * Orthomapper

## ChIP-Seq

 * Read coverage + visualization
 * Differential binding
 * ATAC-Seq

## Multivariate analyses

 * PCA
 * beyond PCA: t-SNE, ICA, UMAP
 * combinding gene set enrichments with PCA

## Single cell RNA-Seq

 * Seurat
 * Visualization
 * Differential analysis
 * Gene set enrichment analysis

# CUBI reports

## CUBI pipelines

 * What is a pipeline?
 * How does it work?
 * What output does it produce?

## Example report

 * GSE147507


## Homework

 * Find GSE147507 on GEO
 * Download counts data
 * Read counts data into R
 * Visualize the counts
